web/docs/internal/keystroke-lifecycle.md

```diff
@@ -0,0 +1,90 @@
+# The Keyman Engine for Web keystroke lifecycle
+
+Depending on the variant of Keyman Engine for Web in play, the engine integrates either with a host app or with the current browser page to interpret and/or intercept keystroke events and produce the desired context manipulations corresponding to them.  While some details diverge between `app/browser` and `app/webview` builds, the general structure and pattern for this process holds.
+
+This document will break down the full keystroke lifecycle into four distinct phases, as follows:
+
+1. Keystroke pre-processing: event interception + interpretation (pre-processing)
+2. Keystroke processing: keystroke main-group rule evaluation
+3. Keystroke post-processing: keystroke post-keystroke evaluation + side-effects
+4. Keystroke resolution: UI state updating
+
+## Keystroke pre-processing
+
+Both the `app/browser` and `app/webview` products are capable of handling simulated keystrokes from the on-screen keyboard as well as hardware keystrokes.  For hardware keystrokes, this varies between the two, while the majority of on-screen keyboard keystrokes are handled identically.
```

Comment:
I think we should avoid the term 'simulated keystroke'. From the perspective of the user, a touch on the their phone (on screen) keyboard is a keystroke -- yes, it's a different kind of keystroke to a traditional hardware keyboard. A simulated keystroke seems like something for a test harness.

So: instead of 'simulated keystroke from the on-screen keyboard' could we say 'touch (or mouse click) on the on-screen keyboard' and instead of 'hardware keystrokes' say 'keypress on the hardware keyboard'. And if we then define 'keystroke' as being either of these, I think we maybe set ourselves up better for future discussion?